docs(lez): Extensions section with admin-authority and extension-library guides (#355)
status-im/ift-docs: docs/lez/extensions/admin-authority.md

:::

`admin-authority` is a SPEL extension that adds a single transferable admin role to your LEZ program. The admin is the only account allowed to call admin-gated instructions. The role can be transferred to another signer or PDA, or renounced permanently. This page walks through using `admin-authority` from an app developer's perspective. If you are building a different extension, see [Build a SPEL extension library](build-a-spel-extension-library.md) instead.

## When to use it

Pick `admin-authority` when your program has:

- A configuration or policy account that only one party should mutate (`set_fee_bps`, `update_oracle_address`, `pause`).
- An emergency action that needs guarded access (`recover_funds`, `migrate_state`).
- A handoff scenario where ownership might rotate between parties over time.

If your program needs multi-party approval rather than single-admin gating, `admin-authority` is the wrong primitive, wait for `multisig-authority` (RFP-TBD) or compose admin-authority with a multisig PDA as the admin.

## Add the dependency

In your program's `Cargo.toml`:

```toml
[dependencies]
admin-authority = { git = "https://github.com/mmlado/spel-admin-authority" }

> spel-framework  = { git = "https://github.com/logos-co/spel" }

Author: Fixed in 992cd40. The snippet pins the v0.1.0 release tag and the framework rev that tag pins, f7aa464. One correction to the suggested fix: branch = "m3" cannot work even when the branch sits on the right commit. Cargo keys git sources by reference kind, so a branch spec never unifies with the rev pin inside the libraries and you get two framework copies and the same trait errors. The pin has to be spelled rev = ..., the same way the libraries spell it. Verified end to end on a clean Ubuntu 24.04 container.

borsh = { version = "1", features = ["derive"] }
serde = { version = "1", features = ["derive"] }
```

All five are needed: the reference samples use exactly this set. `nssa_core` carries the on-chain account types, `borsh` encodes your state, and `serde` is required by the instruction plumbing even when your own types never touch it. The `admin-authority-macros` sub-crate is pulled in transitively. You do not need to declare it directly. The library README documents the framework revision each release is verified against.

After adding the dependencies, run `cargo fetch` once. The framework's extension scanner resolves your dependency graph with an offline metadata call, which fails deterministically for a fresh consumer whose git dependencies were never fetched.

## Annotate the module

If you started from `cargo new`, delete the default `fn main` first. The `#[lez_program]` macro generates the program's entry point, and the leftover stub collides with it as a duplicate `main`.

Add `#[admin_authority]` inside your `#[lez_program]` module:

```rust
use spel_framework::prelude::*;

> use admin_authority::{admin_authority, require_admin};

Author: Fixed in 992cd40. The annotate snippet imports nothing now and the page explains the marker is consumed by the scanner, not resolved as an import. Per your parenthetical, require_admin moved down to the gate section.

```rust
#[instruction]
#[require_admin]
pub fn set_fee_bps(
    #[account(pda = literal("admin_config"))] admin_config: AccountWithMetadata,
    #[account(signer)] caller: AccountWithMetadata,
    #[account(mut, pda = literal("pool_config"))] mut config: AccountWithMetadata,
    new_fee_bps: u16,
) -> SpelResult {
    todo!()
}
```

If your instruction already has parameters by different names, point the gate at them with the inject-account names as keys: `#[require_admin(admin_config = my_cfg, caller = owner)]`. The framework also recognises declared parameters by role, a `#[account(signer)]` parameter or a PDA parameter with the matching seed is reused under its declared name instead of being injected twice.

## Become the first admin

`admin_initialize` takes no arguments. The signing caller becomes the admin (self-election). There is no candidate argument at initialize because the LEZ duplicate-account rule rejects a transaction listing the same account twice, so a caller could never also pass itself as candidate evidence.

```bash

> spel --idl program-idl.json --program <program-id> -- \

Author: Fixed in 992cd40. There is now a Prerequisites section listing what the dependency tree and the CLI build need, unzip and python3-dev included, and an Install the spel CLI section with a cargo install command pinned to the framework rev. The spel vs spel-cli package name is called out.

- **The offset is never in a transaction.** It compiles into the program as a literal. The IDL carries no offset argument, and writing `admin_config = ...` or `offset = ...` on a gate by hand is a compile error in embedded mode.
- **One account fewer** on every gated transaction, the slot travels with state you were already passing.

The embedded `AdminConfig` field must sit at the declared offset with only fixed-size fields before it. The library repository ships a reference sample (`admin-authority-sample-embedded`) with the layout, tests, and a committed dry-run walkthrough.

## Renounce admin permanently

```bash
spel --idl program-idl.json --program <program-id> -- \
    admin-renounce --caller <current-admin-account-id>
```

This writes `AccountId::default()` to the Config PDA. All future admin-gated instructions reject with an authorization error. There is no recovery path, design your program so renounce is only callable when permanent loss of mutability is the intended outcome (handoff to "immutable" governance, end of life, etc.).

## Verify your integration

After building your program, check that the admin instructions appear in the IDL:

```bash

> spel generate-idl path/to/your/program/src/main.rs | jq '.instructions[].name'

Author: Fixed in 57e4d16. The verify section says the CLI must be built from the pinned framework revision and that a scanner-less CLI omits the instructions with a clean exit, and it points at the install section, which pins the right rev.

```

Expected output includes:

```
"admin_initialize"
"admin_transfer"
"admin_renounce"
```

> Plus your own instructions. A marker that matches no discoverable extension is a hard compile error naming the marker, so a broken setup refuses loudly rather than building without the trio. When you hit that error, the most common causes are:

Author: Fixed in 57e4d16. The guarantee is stated as a property of the pinned framework revision and the silent path on upstream main is spelled out.

## When to use it

Pick `admin-authority` when your program has:

- A configuration or policy account that only one party should mutate (`set_fee_bps`, `update_oracle_address`, `pause`).
- An emergency action that needs guarded access (`recover_funds`, `migrate_state`).
- A handoff scenario where ownership might rotate between parties over time.

If your program needs multi-party approval rather than single-admin gating, `admin-authority` is the wrong primitive, wait for `multisig-authority` (RFP-TBD) or compose admin-authority with a multisig PDA as the admin.

## Prerequisites

You need a stable Rust toolchain, git, and the native build tools the dependency tree leans on. The `spel` CLI additionally needs `unzip` (a build script unpacks a prebuilt rapidsnark archive) and the Python development library (the CLI links against libpython). The verification step at the end uses `jq`. On a fresh Ubuntu 24.04 this covers everything:

```bash

> sudo apt-get install curl git build-essential pkg-config libssl-dev ca-certificates unzip python3 python3-dev cmake jq

Reviewer: This line does not work on the clean Ubuntu 24.04 the sentence above promises it covers. A base `ubuntu:24.04` image ships with `/var/lib/apt/lists` empty, so `apt-get install` has no package index to resolve against:

```
$ apt-get install -s --no-install-recommends <pkg>     # with empty apt lists
Reading package lists...
Building dependency tree...
Reading state information...
E: Unable to locate package <pkg>
```

`apt-get update` first fixes it. I also added `-y`, since a reader pasting this into a container script otherwise stalls on the confirmation prompt.

```suggestion
sudo apt-get update && sudo apt-get install -y curl git build-essential pkg-config libssl-dev ca-certificates unzip python3 python3-dev cmake jq
```

```

> The build and IDL verification steps on this page were verified on a clean Ubuntu 24.04 with rustc 1.98. The lifecycle commands have not been run against a live node yet.

Reviewer: A spot version is harder to act on than a floor, and it reads as a requirement. I ran every build and IDL step on this page — dedicated mode and embedded mode — on **rustc 1.94.1** and they all pass, so 1.98 is not a floor. The real constraint comes from the libraries: `admin-authority` declares `rust-version = "1.88"` and the workspace is `edition = "2024"` (which itself needs ≥ 1.85).

```suggestion
The build and IDL verification steps on this page need a toolchain new enough for edition 2024; `admin-authority` declares `rust-version = "1.88"`. They were verified on a clean Ubuntu 24.04 with rustc 1.94.1 and 1.98. The lifecycle commands have not been run against a live node yet.
```

The `spel-framework` entry points at a fork on purpose. It must be the exact revision `admin-authority` itself pins, and the library README documents that revision for each release. Pointing at `logos-co/spel` instead puts two copies of the framework into your dependency graph, and the build fails with a `From<AdminError>` trait error plus name resolution errors inside the `require_admin` expansion. The dependency moves to `logos-co/spel` once the extension mechanism lands upstream ([logos-co/spel#257](https://github.com/logos-co/spel/pull/257)).

After adding the dependencies, run `cargo fetch` once. The framework's extension scanner resolves your dependency graph with an offline metadata call, which fails deterministically for a fresh consumer whose git dependencies were never fetched.

## Install the spel CLI

The lifecycle commands below and the IDL check at the end use the `spel` CLI. Install it from the same fork revision the framework dependency pins:

```bash
cargo install --git https://github.com/mmlado/spel --rev f7aa464b2c6c72ef513a25ede16584bca85b722f spel
```

> The package name is `spel`, not `spel-cli` as the repository directory suggests, asking cargo for `spel-cli` fails with "could not find `spel-cli`".

Reviewer: The `spel-cli` claim is exactly right — I checked it:

```
$ cargo install --git https://github.com/mmlado/spel --rev f7aa464... spel-cli
error: could not find `spel-cli` in https://github.com/mmlado/spel?rev=f7aa464... with version `*`
```

Two additions to this line.

First, one thing the Prerequisites section does not warn about cost me a failed install: this `cargo install` reaches the network *during the build*, beyond fetching crates. A transitive build script (`logos-blockchain-circuits-pol-sys`) downloads a prebuilt circuits artifact from GitHub releases with `ureq`. `ureq` does not read the usual CA environment variables, so behind any TLS-inspecting proxy the install dies well into a long release build:

```
Failed to download a prebuilt artifact for linux-x86_64 v0.5.3:
io: invalid peer certificate: UnknownIssuer. Set LBC_ROOT_DIR to point to a local build instead.
```

I got past it by fetching the tarball myself and unpacking it into `~/.cache/logos/blockchain/`, after which the documented command succeeded unchanged. Worth a sentence, since the failure appears minutes in and the escape hatch is not discoverable from the error alone.

Second, `Google.Quotes` flags the trailing period here (see the Vale note in the review summary); moving it inside the quotes clears it.

```suggestion
The package name is `spel`, not `spel-cli` as the repository directory suggests, asking cargo for `spel-cli` fails with "could not find `spel-cli`."

This build also reaches the network beyond fetching crates: a transitive build script downloads a prebuilt `logos-blockchain-circuits` artifact from GitHub releases. It does not honour the usual CA environment variables, so behind a TLS-inspecting proxy it fails with `invalid peer certificate: UnknownIssuer` partway through the build. Set `LBC_ROOT_DIR` to a local circuits build, or unpack the release tarball into `~/.cache/logos/blockchain/` beforehand, if the download cannot reach GitHub directly.
```

Add `#[admin_authority]` inside your `#[lez_program]` module:

```rust
use spel_framework::prelude::*;

#[lez_program]
#[admin_authority]
mod my_program {
    #[instruction]
    pub fn create_pool(
        #[account(init, pda = literal("pool"))] pool: AccountWithMetadata,
    ) -> SpelResult { /* ... */ }
}
```

Nothing is imported from the library at this point. The `#[admin_authority]` marker is consumed by the framework's scanner during expansion, not resolved as an import, so importing the name only earns an unused import warning. The gate attribute gets imported when the first instruction uses it, next section. The module body does not need `use super::*;` either, the macro resolves paths to items declared outside the module on its own.

That single annotation exposes three new instructions in your program's IDL:

| Instruction | Purpose |
|---|---|
| `admin_initialize` | Creates the admin Config PDA and installs the caller as the first admin. Must be called once after deployment. |
| `admin_transfer` | Replaces the current admin with a new signer or PDA. |
| `admin_renounce` | Zeros the admin permanently. Terminal, no recovery path. |

:::warning

> **Initialization window.** Until `admin_initialize` is called, the admin Config PDA does not exist. Anyone who submits the first `admin_initialize` becomes the admin. Send it as the very next transaction after deployment to prevent a third party from claiming the role. Bundling with the deployment itself is not possible today because a LEZ deployment transaction carries no instructions.

Reviewer: The Vale check fails on this branch (details and the full breakdown are in the review summary). This is the first of the British-spelling errors; `Logos.SpellingGB` is `level: error` and the workflow runs with `fail_on_error: true`.

```suggestion
**Initialisation window.** Until `admin_initialize` is called, the admin Config PDA does not exist. Anyone who submits the first `admin_initialize` becomes the admin. Send it as the very next transaction after deployment to prevent a third party from claiming the role. Bundling with the deployment itself is not possible today because a LEZ deployment transaction carries no instructions.
```

The same class of fix is needed at these lines (identifiers inside backticks are exempt — Vale skips code spans, so `admin_initialize` and friends stay as they are):

| File | Line | Change |
|---|---|---|
| `admin-authority.md` | 154, 161 | initialize → initialise |
| `admin-authority.md` | 204, 264 | authorization → authorisation |
| `admin-authority.md` | 249 | initialized → initialised, initialization → initialisation |
| `admin-authority.md` | 251 | neighboring → neighbouring |
| `admin-authority.md` | 292 | Initialization → Initialisation |
| `build-a-spel-extension-library.md` | 206 | synthesized → synthesised |
| `build-a-spel-extension-library.md` | 260 | uninitialized → uninitialised, synthesize → synthesise |
| `build-a-spel-extension-library.md` | 284 | initialization → initialisation |
| `build-a-spel-extension-library.md` | 299 | spelled → spelt |
| `freeze-authority.md` | 56 | spelled → spelt |
| `freeze-authority.md` | 121 | Initialization → Initialisation, initialization → initialisation |
| `freeze-authority.md` | 159 | Initialize → Initialise |
| `freeze-authority.md` | 311 | initialization → initialisation |
| `freeze-authority.md` | 314 | neighboring → neighbouring |
| `freeze-authority.md` | 353 | Initialization → Initialisation, initialized → initialised |
| `freeze-authority.md` | 357 | initialized → initialised |

This matches the rest of the repo, which is consistently British in prose — 49 `initialise` against 3 `initialize` across `docs/`, and every other page under `docs/lez/` currently reports zero Vale errors.

```bash
spel --idl program-idl.json --program <program-id> -- \
    admin-initialize --caller <your-account-id>
```

To hand the role to a different keyholder or a PDA, initialize first and then call `admin_transfer`.

## Transfer admin to another party

`admin_transfer` requires the current admin to sign. It takes an `AdminCandidate` describing the new admin, paired with a matching `AccountWithMetadata` that proves the candidate on chain.

Two candidate shapes:

```rust
pub enum AdminCandidate {
    /// The new admin is a keyholder. Validated by checking the new account
    /// co-signed the transaction.
    Signer,
    /// The new admin is a program-owned PDA. Validated by deriving the address
    /// from (program_id, seed) and confirming the PDA exists on chain.
    Pda { program_id: ProgramId, seed: [u8; 32] },
}
```

```bash
spel --idl program-idl.json --program <program-id> -- \
    admin-transfer \
    --caller <current-admin-account-id> \
    --new-account <new-admin-account-id> \
    --candidate Signer
```

> A `Signer` transfer needs the new admin's signature on the same transaction, which proves the keyholder consents. That means two parties sign one message, an off-chain co-signing exchange: export the partial transaction with the candidate named as a co-signer (`--export handover.json --co-signer <new-admin-account-id-hex>`), send the file to the candidate to run `spel sign`, then submit it. Without the second signature the sequencer drops the transaction.

Reviewer: **Blocker — `--export`, `--co-signer` and `spel sign` do not exist in the CLI this page pins.**

I followed the page end to end on a clean Ubuntu 24.04, built the CLI from the revision the [Install the spel CLI](#install-the-spel-cli) section pins (`f7aa464`), and this handover flow cannot be carried out as written.

What I ran and what happened:

```
$ spel --idl pool-idl.json --program <pid> -- admin-transfer \
    --caller <admin> --new-account <new> --candidate Signer \
    --export handover.json --co-signer <new>
...
   Set LEE_WALLET_HOME_DIR environment variable   # panicked on the submit path
$ ls handover.json
ls: cannot access 'handover.json': No such file or directory
```

The run is byte-for-byte identical to the run with no extra flags at all, because the CLI silently ignores unrecognised flags. So a reader following this paragraph believes they exported a partial transaction for counter-signature, while the CLI actually went straight to submitting. That is worse than a missing feature.

Confirming it is not a spelling difference on my side:

```
$ strings ~/.cargo/bin/spel | grep -cE 'co-signer|--export'
0
$ grep -rniE '"--export"|co.signer|co_signer|"sign"' --include=*.rs ~/.cargo/git/checkouts/spel-*/f7aa464/spel-cli/
(no matches)
```

`spel sign` is not a subcommand either — `spel sign --help` just prints the top-level usage. The only subcommands at `f7aa464` are `init`, `program-id`, `inspect`, `generate-idl` plus the IDL-derived instruction commands; the only global options are `--idl`, `--program`, `--program-id`, `--dry-run`, `--format` and `--bin-<NAME>`.

Everything else about the transfer section checks out — `admin-transfer --caller --new-account --candidate` is exactly right, and `--candidate Signer` resolves correctly under `--dry-run`. It is only the co-signing exchange that is not implemented yet, so I suggest stating that rather than describing an interface that is not there:

```suggestion
A `Signer` transfer needs the new admin's signature on the same transaction, which proves the keyholder consents. That means two parties sign one message. The `spel` CLI at the pinned revision has no co-signing exchange: the command above builds and submits with the caller's signature only, and the sequencer drops the transaction unless the candidate's signature is attached. Collecting that second signature is not yet possible from `spel` itself.
```

:::

`admin-authority` is a SPEL extension that adds a single transferable admin role to your LEZ program. The admin is the only account allowed to call admin-gated instructions. The role can be transferred to another signer or PDA, or renounced permanently. This page walks through using `admin-authority` from an app developer's perspective. If you are building a different extension, see [Build a SPEL extension library](build-a-spel-extension-library.md) instead.

## When to use it

Pick `admin-authority` when your program has:

- A configuration or policy account that only one party should mutate (`set_fee_bps`, `update_oracle_address`, `pause`).
- An emergency action that needs guarded access (`recover_funds`, `migrate_state`).
- A handoff scenario where ownership might rotate between parties over time.

If your program needs multi-party approval rather than single-admin gating, `admin-authority` is the wrong primitive, wait for `multisig-authority` (RFP-TBD) or compose admin-authority with a multisig PDA as the admin.

## Prerequisites

> You need a stable Rust toolchain, git, and the native build tools the dependency tree leans on. The `spel` CLI additionally needs `unzip` (a build script unpacks a prebuilt rapidsnark archive) and the Python development library (the CLI links against libpython). The verification step at the end uses `jq`. On a fresh Ubuntu 24.04 this covers everything:

Author: unzip is needed. It isn't in any .rs file, it is in a shell script that build.rs embeds and executes. rust-rapidsnark v0.1.3 at the rev this tree resolves (e91187f8, pulled in through the circuits prover):

- [crates/build.rs](https://github.com/logos-blockchain/logos-blockchain-rust-rapidsnark/blob/e91187f8ccb5bbfc7bb00dac88169112428da78f/crates/build.rs#L79-L82) includes download_rapidsnark.sh as a string (line 6) and, when RAPIDSNARK_LIB_DIR is unset (lines 21-25), writes it to OUT_DIR and runs it with sh.
- [crates/download_rapidsnark.sh](https://github.com/logos-blockchain/logos-blockchain-rust-rapidsnark/blob/e91187f8ccb5bbfc7bb00dac88169112428da78f/crates/download_rapidsnark.sh#L54-L60) curls rapidsnark-linux-x86_64-pic-v0.0.8.zip (line 54) and unpacks it with unzip -o (line 60).
The tar.gz you found is the separate logos-blockchain-circuits artifact, which really is unpacked in-process.

I re-verified differentially on a clean ubuntu:24.04: without unzip, cmake present, cargo install of spel at f7aa464 fails with "Failed to unzip .../rapidsnark-linux-x86_64-pic-v0.0.8.zip". Installing unzip into the same container as the only change makes the identical command succeed. Without it: the build script skips the whole download when it finds an existing unpacked artifact for the arch (the check around build.rs line 99), so a warm build directory, or RAPIDSNARK_LIB_DIR in the environment, never reaches the unzip call. The prerequisites sentence now names the mechanism for both packages so the next reader can verify it the same way.

```bash

> sudo apt-get update && sudo apt-get install -y curl git build-essential pkg-config libssl-dev ca-certificates unzip python3 python3-dev cmake jq

Author: Kept as is apart from earlier fixes, see the thread above, unzip turned out to be neededl. The update and -y parts stay, and thank you for confirming cmake and python3-dev are load-bearing.

```

> The build and IDL verification steps on this page need a toolchain new enough for edition 2024, `admin-authority` declares `rust-version = "1.88"`. They were verified on a clean Ubuntu 24.04 with rustc 1.94.1 and 1.98. The lifecycle commands were verified against a live LEZ stack during the library's milestone reviews, on the same framework revision this page pins. The co-signing exchange is the one exception, see the transfer section.

Author: You are right, and the 1.88.0 resolver output settles it. Applied your wording: the floor is 1.94.1, it comes from the wider tree, and the sentence explains where admin-authority's own 1.88 fits.

After adding the dependencies, run `cargo fetch` once. The framework's extension scanner resolves your dependency graph with an offline metadata call, which fails deterministically for a fresh consumer whose git dependencies were never fetched.

## Install the `spel` CLI

The lifecycle commands below and the IDL check at the end use the `spel` CLI. Install it from the same fork revision the framework dependency pins:

```bash
cargo install --git https://github.com/mmlado/spel --rev f7aa464b2c6c72ef513a25ede16584bca85b722f spel
```

The package name is `spel`, not `spel-cli` as the repository directory suggests, asking cargo for `spel-cli` fails with "could not find `spel-cli`."

> This build also reaches the network beyond fetching crates: a transitive build script downloads a prebuilt `logos-blockchain-circuits` artifact from GitHub releases. It does not honour the usual CA environment variables, so behind a TLS-inspecting proxy it fails with `invalid peer certificate: UnknownIssuer` partway through the build. Set `LBC_ROOT_DIR` to a local circuits build, or unpack the release tarball into `~/.cache/logos/blockchain/` beforehand, if the download cannot reach GitHub directly.

Author: Applied with your pre-seed command and the version-match note. Good catch, that the tarball's top-level entry already matches the directory the build script asserts on.

## Become the first admin

`admin_initialize` takes no arguments. The signing caller becomes the admin (self-election). There is no candidate argument at initialise because the LEZ duplicate-account rule rejects a transaction listing the same account twice, so a caller could never also pass itself as candidate evidence.

```bash
spel --idl program-idl.json --program <program-id> -- \
    admin-initialize --caller <your-account-id>
```

To hand the role to a different keyholder or a PDA, initialise first and then call `admin_transfer`.

## Transfer admin to another party

`admin_transfer` requires the current admin to sign. It takes an `AdminCandidate` describing the new admin, paired with a matching `AccountWithMetadata` that proves the candidate on chain.

Two candidate shapes:

```rust
pub enum AdminCandidate {
    /// The new admin is a keyholder. Validated by checking the new account
    /// co-signed the transaction.
    Signer,
    /// The new admin is a program-owned PDA. Validated by deriving the address
    /// from (program_id, seed) and confirming the PDA exists on chain.
    Pda { program_id: ProgramId, seed: [u8; 32] },

> }

Author: Applied. The snippet now shows the alias comment and the real enum name, and the freeze page says the two candidates are the same type rather than the same shape.

## When to use it

Pick `admin-authority` when your program has:

- A configuration or policy account that only one party should mutate (`set_fee_bps`, `update_oracle_address`, `pause`).
- An emergency action that needs guarded access (`recover_funds`, `migrate_state`).
- A handoff scenario where ownership might rotate between parties over time.

If your program needs multi-party approval rather than single-admin gating, `admin-authority` is the wrong primitive, wait for `multisig-authority` (RFP-TBD) or compose admin-authority with a multisig PDA as the admin.

## Prerequisites

You need a stable Rust toolchain, git, and the native build tools the dependency tree leans on. The `spel` CLI additionally needs `unzip` (the `rust-rapidsnark` build script downloads a prebuilt zip and unpacks it through a helper shell script) and the Python development library (a transitive dependency links against libpython through `pyo3`). The verification step at the end uses `jq`. On a fresh Ubuntu 24.04 this covers everything:

```bash
sudo apt-get update && sudo apt-get install -y curl git build-essential pkg-config libssl-dev ca-certificates unzip python3 python3-dev cmake jq
```

> The build and IDL verification steps on this page need rustc 1.94.1 or newer. That floor comes from the wider dependency tree rather than from `admin-authority` itself, which declares only `rust-version = "1.88"`: the `spel` CLI pins `channel = "1.94.0"` in its own `rust-toolchain.toml`, and transitive crates in the program's dependency graph already require 1.90. They were verified on a clean Ubuntu 24.04 with rustc 1.94.1 and 1.98. The lifecycle commands were verified against a live LEZ stack during the library's milestone reviews, on the same framework revision this page pins. The co-signing exchange is the one exception, see the transfer section.

Reviewer: **The stated floor is wrong in both directions, and it merges two genuinely different floors into one number.**

I tested this one directly rather than reading the manifests, because the sentence's own evidence doesn't add up to `1.94.1`: the CLI pins `1.94.0`, and the transitive crates are `1.90`. Neither of those is `1.94.1` — that was just the version I happened to run last time.

Building the program does **not** need 1.94.1. It builds clean on 1.90.0:

```
$ cargo +1.90.0 build
   Compiling admin-authority v0.1.0 (...tag=v0.1.0#34574371)
   Compiling admin-consumer v0.1.0
    Finished `dev` profile in 1m 08s
```

That matches the graph: the highest `rust-version` anywhere in the program's resolved dependencies is `1.90`, from `ruint 1.20.0`.

Installing the CLI **does** need a newer toolchain, but the floor is `1.94.0`, not `1.94.1` — 1.94.0 builds and runs it:

```
$ cargo +1.94.0 install --git https://github.com/mmlado/spel --rev f7aa464... spel --root /tmp/spel1940
    Installed package `spel v0.6.0`
$ /tmp/spel1940/bin/spel generate-idl admin-consumer/src/main.rs | jq -r '.instructions[].name'
create_pool
set_fee_bps
admin_initialize
admin_transfer
admin_renounce
```

and 1.90.0 fails, which is where that floor actually comes from:

```
$ cargo +1.90.0 install --git https://github.com/mmlado/spel --rev f7aa464... spel
error[E0658]: use of unstable library feature `strict_overflow_ops`
error: could not compile `logos-blockchain-cryptarchia-engine` (lib) due to 6 previous errors
```

That failure is also worth stating explicitly in the page, because it shows the `rust-toolchain.toml` pin is **not** doing the work the sentence implies: if `cargo install --git` applied the dependency's toolchain file, rustup would have switched to 1.94.0 and this would have succeeded. It doesn't, so the reader's own default toolchain has to clear the bar.

```suggestion
Two different floors apply here, and neither comes from `admin-authority` itself, which declares only `rust-version = "1.88"`. Building your program needs rustc 1.90 or newer, a floor set by transitive crates in its dependency graph (`ruint 1.20.0`). Installing the `spel` CLI needs rustc 1.94.0 or newer: a transitive `logos-blockchain` crate uses `strict_overflow_ops`, which older toolchains reject as an unstable feature. The CLI pins `channel = "1.94.0"` in its own `rust-toolchain.toml`, but `cargo install --git` does not apply a dependency's toolchain file, so your default toolchain has to meet that floor itself. Verified on a clean Ubuntu 24.04: the program builds on rustc 1.90.0 and 1.94.1, and the CLI installs on 1.94.0, 1.94.1 and 1.98. The lifecycle commands were verified against a live LEZ stack during the library's milestone reviews, on the same framework revision this page pins. The co-signing exchange is the one exception, see the transfer section.
```

---
_Generated by [Claude Code](https://claude.ai/code)_

Author: Applied with your wording and your numbers, then re-verified independently: the program builds on 1.90.0 with a fresh target directory, 1.88 refuses at resolution with the exact enum-ordinalize and ruint lines you quoted, and the CLI installs on 1.94.0 and runs. Your cargo install --git point stays in the sentence, that was the part a reader could never have derived from the manifests. The verified list is trimmed to the two floors, everything above them is implied by "or newer".

:::

## Gate an instruction

Add `#[require_admin]` to any instruction that should only succeed when the caller is the current admin:

```rust
#[account_type]
#[derive(BorshSerialize, BorshDeserialize, Clone, Debug)]
pub struct PoolConfig {
    pub fee_bps: u16,
}

// ... inside the #[lez_program] module:
use admin_authority::require_admin;

#[instruction]
#[require_admin]
pub fn set_fee_bps(
    #[account(mut, pda = literal("pool_config"))] mut config: AccountWithMetadata,
    new_fee_bps: u16,
) -> SpelResult {
    // The admin check has already run.
    PoolConfig { fee_bps: new_fee_bps }.write_to(&mut config)?;
    Ok(SpelOutput::execute(vec![config], vec![]))
}
```

The `write_to` helper is yours to write, the library does not provide it. The reference sample uses this one:

```rust
impl PoolConfig {
    fn write_to(&self, account: &mut AccountWithMetadata) -> Result<(), SpelError> {
        account.account.data = borsh::to_vec(self)
            .map_err(|_| SpelError::SerializationError { message: "encoding failed".into() })?
            .try_into()
            .map_err(|_| SpelError::SerializationError { message: "data too large".into() })?;
        Ok(())
    }
}
```

> The `#[account_type]` struct sits outside the `#[lez_program]` module, the instruction inside it. The handler returns `Ok(SpelOutput::execute(post_states, messages))`, where `post_states` lists your declared accounts in declaration order. The injected `admin_config` and `caller` are appended to the post-states automatically, you only handle the parameters you wrote.

Reviewer: **`appended` is the wrong direction — the injected accounts go in front of yours.**

I read this off the actual expansion rather than the IDL, so there's no ambiguity about what the handler receives. `cargo rustc -- -Zunpretty=expanded` on the program built from this page's own snippets:

```rust
pub fn set_fee_bps(admin_config: AccountWithMetadata,
    caller: AccountWithMetadata, mut config: AccountWithMetadata,
    new_fee_bps: u16) -> SpelResult {
    { /* the injected admin check */ }
    PoolConfig { fee_bps: new_fee_bps }.write_to(&mut config)?;
    Ok(SpelOutput::execute_with_claims(&[admin_config.account.clone(),
                        caller.account.clone(), config.account.clone()],
            &__claims_set_fee_bps(), ::alloc::vec::Vec::new()))
}
```

The `vec![config]` I wrote becomes `[admin_config, caller, config]` — the two injected accounts are prepended and my declared account lands last. The IDL agrees:

```
$ spel generate-idl src/main.rs | jq -c '.instructions[]|select(.name=="set_fee_bps")|[.accounts[].name]'
["admin_config","caller","config"]
```

This also contradicts the author guide, which has it right: `build-a-spel-extension-library.md` L206 says injected parameters "land after a leading `ProgramContext` in the block's declaration order", i.e. ahead of the consumer's own. Worth aligning the two pages, since a reader who believes "appended" will get the account order backwards the moment they declare the gate accounts by hand (see the next comment).

```suggestion
The `#[account_type]` struct sits outside the `#[lez_program]` module, the instruction inside it. The handler returns `Ok(SpelOutput::execute(post_states, messages))`, where `post_states` lists your declared accounts in declaration order. The injected `admin_config` and `caller` are prepended to the post-states automatically, ahead of the accounts you declared, you only handle the parameters you wrote.
```

---
_Generated by [Claude Code](https://claude.ai/code)_

Author: Applied.

> The gate needs two accounts, the `admin_config` PDA holding the current admin state and a signing `caller`. You do not have to write them: the framework injects both from metadata the library declares, and they appear in the IDL like declared parameters. Declaring them explicitly produces the same program, and then they are your parameters, appearing in your post-states list like any other account.

Reviewer: **"Declaring them explicitly produces the same program" only holds if you declare them first, in the injected order.**

This is the failure the previous comment sets up, so I tested it rather than reasoning about it. Two programs, identical except for where the gate accounts are declared.

Declared the way a reader naturally writes them — own parameter first, gate accounts after:

```rust
pub fn set_fee_bps(
    #[account(mut, pda = literal("pool_config"))] mut config: AccountWithMetadata,
    #[account(pda = literal("admin_config"))] admin_config: AccountWithMetadata,
    #[account(signer)] caller: AccountWithMetadata,
    new_fee_bps: u16,
) -> SpelResult { ... }
```

```
injected : ["admin_config","caller","config"]
explicit : ["config","admin_config","caller"]
```

Both compile and both build a working program, but the account order differs, and that order is the transaction ABI — the same property `build-a-spel-extension-library.md` L339 calls "the cross-extension ABI". A caller built against one account order does not work against the other.

Declaring them ahead of my own parameter, in the injected order, does give a genuinely identical program — I diffed the two IDL instruction objects and they match exactly:

```
$ diff <(spel generate-idl ../admin-consumer/src/main.rs | jq -S '.instructions[]|select(.name=="set_fee_bps")') \
       <(spel generate-idl src/main.rs          | jq -S '.instructions[]|select(.name=="set_fee_bps")')
$ echo $?
0
```

So the claim is true, just not unconditionally, and the condition is the part a reader needs.

```suggestion
The gate needs two accounts, the `admin_config` PDA holding the current admin state and a signing `caller`. You do not have to write them: the framework injects both from metadata the library declares, and they appear in the IDL like declared parameters. Declaring them explicitly produces the same program only if you declare them ahead of your own parameters and in the injected order, `admin_config` then `caller`; they are then your parameters, appearing in your post-states list like any other account. Declaring them after your own parameters still compiles, but it reorders the instruction's accounts in the IDL, and that order is the transaction ABI.
```

---
_Generated by [Claude Code](https://claude.ai/code)_

Author: Applied with your condition and the ABI consequence spelt out.

If your instruction already has parameters by different names, point the gate at them with the inject-account names as keys: `#[require_admin(admin_config = my_cfg, caller = owner)]`. The framework also recognises declared parameters by role, a `#[account(signer)]` parameter or a PDA parameter with the matching seed is reused under its declared name instead of being injected twice.

## Become the first admin

> `admin_initialize` takes no arguments. The signing caller becomes the admin (self-election). There is no candidate argument at initialise because the LEZ duplicate-account rule rejects a transaction listing the same account twice, so a caller could never also pass itself as candidate evidence.

Reviewer: **Blocker — the page never creates `program-idl.json`, and this is the first command that needs it.**

I followed the page top to bottom on a clean Ubuntu 24.04: prerequisites, dependency block, `cargo fetch`, `cargo build`, `cargo install ... spel`. All of that works. Then this command is the first to read `--idl program-idl.json`, and nothing above ever wrote that file:

```
$ spel --idl program-idl.json --program <pid> -- admin-initialize --caller <acct>
Error reading IDL 'program-idl.json': No such file or directory (os error 2)
```

`program-idl.json` is referenced by **12 command blocks** across the two pages (`admin-authority.md` L167, L193, L209, L272; `freeze-authority.md` L164, L172, L176, L186, L190, L207, L221, L234) and produced by none of them. The only `generate-idl` invocation on either page is in [Verify your integration](#verify-your-integration) — 120 lines *below* this one — and it pipes to `jq`, so it discards the IDL rather than saving it.

The fix is one line, and it works from any directory:

```
$ spel generate-idl admin-consumer/src/main.rs > program-idl.json
$ jq -r '.instructions[].name' program-idl.json
create_pool
set_fee_bps
admin_initialize
admin_transfer
admin_renounce
```

I'd also mention `--dry-run` here. The page says the lifecycle commands were verified against a live stack but that the co-signing exchange is missing from the pinned CLI, which leaves a reader with no way to check their own invocation. `--dry-run=text` resolves and prints the transaction without submitting, and it validates the whole command including the candidate payloads further down:

```
$ spel --idl program-idl.json --program <pid> --dry-run=text -- \
    admin-transfer --caller <admin> --new-account <new> --candidate Signer
  caller → 0x2222...2222 [signer]
  new_account → 0x3333...3333
Arguments:
  --candidate Signer
Instruction data: 0x0600000000000000
Dry run complete — not submitted.
```

Note `--dry-run` has to sit before the `--` separator and needs its value spelt with `=`; `-- ... --dry-run` at the end fails with `❌ --dry-run: missing value`.

The same gap exists on `freeze-authority.md` at [Initialise the freeze authority](freeze-authority.md#initialise-the-freeze-authority) (L163). That page defers its toolchain setup to this one, so a sentence there pointing back at this section would cover it.

````suggestion
`admin_initialize` takes no arguments. The signing caller becomes the admin (self-election). There is no candidate argument at initialise because the LEZ duplicate-account rule rejects a transaction listing the same account twice, so a caller could never also pass itself as candidate evidence.

Every lifecycle command below reads your program's IDL from a file. Generate it once from the program you just built:

```bash
spel generate-idl path/to/your/program/src/main.rs > program-idl.json
```

To check a command without submitting it, add `--dry-run=text` before the `--` separator. It resolves the accounts and arguments and prints the transaction it would send, which is the only way to validate an invocation before you have a node to submit to.
````

---
_Generated by [Claude Code](https://claude.ai/code)_

Author: Applied: the IDL is generated once right after the build, before the first command that reads it, and the --dry-run=text paragraph is in, = spelling and placement before the separator included. The freeze page points back at this section from its initialise step.
